crates/host/src/module.rs

```diff
@@ -266,109 +231,99 @@ impl SerializedModuleCache {
                 // instantiated with fresh stores free from state. Instance
                 // creation is highly performant which makes caching of instances
                 // and stores unnecessary.
-                let module = unsafe {
+                //
+                // See https://github.com/wasmerio/wasmer/issues/4377
+                let serialized_module = module
+                    .serialize()
+                    .map_err(|e| wasm_error!(WasmErrorInner::ModuleBuild(e.to_string())))?;
+                let module = Arc::new(unsafe {
                     Module::deserialize(&self.runtime_engine, serialized_module.clone())
                         .map_err(|e| wasm_error!(WasmErrorInner::ModuleBuild(e.to_string())))?
-                };
+                });
 
-                (module, serialized_module)
-            }
-        };
-        self.put_item(key, Arc::new(serialized_module.clone()));
+                // Save serialized module to filesystem cache
+                self.add_to_filesystem_cache(key, serialized_module)?;
```

Comment:
Note that we *first* save to filesystem, *then* add to in-memory cache. Thus if the process is interrupted after saving to the filesystem, it will still be added to the in-memory cache next time it is fetched.

This also made me think of an issue -- if the file is corrupted then it will error every time and never be purged from the cache. I made an issue to address this: https://github.com/holochain/holochain-wasmer/issues/137